hudi-client/hudi-flink-client/src/main/java/org/apache/hudi/metadata/FlinkHoodieBackedTableMetadataWriter.java

```diff
@@ -97,68 +96,72 @@ protected void initRegistry() {
+      // when retried again, data table will first rollback pending compaction. these will be applied to metadata table, but all changes
+      // are upserts to metadata table and so only a new delta commit will be created.
+      // once rollback is complete in datatable, compaction will be retried again, which will eventually hit this code block where the respective commit is
+      // already part of completed commit. So, we have to manually rollback the completed instant and proceed.
+      Option<HoodieInstant> alreadyCompletedInstant = metadataMetaClient.getActiveTimeline().filterCompletedInstants().filter(entry -> entry.getTimestamp().equals(instantTime))
+          .lastInstant();
+      LOG.info(String.format("%s completed commit at %s being applied to MDT.",
+          alreadyCompletedInstant.isPresent() ? "Already" : "Partially", instantTime));
+
+      // Rollback the previous commit
+      if (!writeClient.rollback(instantTime)) {
+        throw new HoodieMetadataException("Failed to rollback deltacommit at " + instantTime + " from MDT");
       }
+      metadataMetaClient.reloadActiveTimeline();
+    }
 
-      writeClient.startCommitWithTime(instantTime);
-      metadataMetaClient.getActiveTimeline().transitionRequestedToInflight(HoodieActiveTimeline.DELTA_COMMIT_ACTION, instantTime);
+    writeClient.startCommitWithTime(instantTime);
+    preWrite(instantTime);
 
-      List<WriteStatus> statuses = isInitializing
-          ? writeClient.bulkInsertPreppedRecords(preppedRecordList, instantTime, Option.empty())
-          : writeClient.upsertPreppedRecords(preppedRecordList, instantTime);
-      // flink does not support auto-commit yet, also the auto commit logic is not complete as BaseHoodieWriteClient now.
-      writeClient.commit(instantTime, statuses, Option.empty(), HoodieActiveTimeline.DELTA_COMMIT_ACTION, Collections.emptyMap());
+    List<WriteStatus> statuses = isInitializing
+        ? writeClient.bulkInsertPreppedRecords(preppedRecordList, instantTime, Option.empty())
+        : writeClient.upsertPreppedRecords(preppedRecordList, instantTime);
+    // flink does not support auto-commit yet, also the auto commit logic is not complete as BaseHoodieWriteClient now.
+    writeClient.commit(instantTime, statuses, Option.empty(), HoodieActiveTimeline.DELTA_COMMIT_ACTION, Collections.emptyMap());
 
-      // reload timeline
-      metadataMetaClient.reloadActiveTimeline();
-      metadataMetaClient.getActiveTimeline().getDeltaCommitTimeline().filterCompletedInstants().lastInstant().ifPresent(instant -> cleanIfNecessary(writeClient, instant.getTimestamp()));
-      writeClient.archive();
-    }
+    // reload timeline
+    metadataMetaClient.reloadActiveTimeline();
+    metadataMetaClient.getActiveTimeline().getDeltaCommitTimeline().filterCompletedInstants().lastInstant().ifPresent(instant -> cleanIfNecessary(writeClient, instant.getTimestamp()));
+    writeClient.archive();
```

Comment:
I'm wondering when the write client got closed properly?

Reply:
within the close method.

Reply:
Okay, seems the write client is shared by multiple MDT writer methods and has the same lifecycle with the writer itself now.